From 1211bb3315c369c2a21acffaf5fc0a1fb42d12a7 Mon Sep 17 00:00:00 2001
From: Dmitri Epshtein <dima@marvell.com>
Date: Mon, 11 Nov 2013 10:09:11 -0500
Subject: [PATCH 1068/1825] fix: pp2: Fix Interrupt mask/unmask processing.

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit a9555a6a88ffbd3e79de0f577d1762df88d358e6

	Prevent wrong enable interrupts for non-initialized ports

Change-Id: I50989966a74e3ebcaf3b64990f83ffa8df3821ca
Signed-off-by: Dmitri Epshtein <dima@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/4276
Reviewed-by: Yelena Krivosheev <yelena@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Nadav Haklai <nadavh@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 .../mv_drivers_lsp/mv_pp2/net_dev/mv_ethernet.c    |    7 +++++++
 .../mv_drivers_lsp/mv_pp2/net_dev/mv_netdev.c      |   17 -----------------
 2 files changed, 7 insertions(+), 17 deletions(-)

diff --git a/arch/arm/plat-armada/mv_drivers_lsp/mv_pp2/net_dev/mv_ethernet.c b/arch/arm/plat-armada/mv_drivers_lsp/mv_pp2/net_dev/mv_ethernet.c
index d09562e..8729e69 100644
--- a/arch/arm/plat-armada/mv_drivers_lsp/mv_pp2/net_dev/mv_ethernet.c
+++ b/arch/arm/plat-armada/mv_drivers_lsp/mv_pp2/net_dev/mv_ethernet.c
@@ -101,6 +101,9 @@ int mv_eth_start(struct net_device *dev)
 		mv_eth_interrupts_unmask(priv);
 		smp_call_function_many(cpu_online_mask, (smp_call_func_t)mv_eth_interrupts_unmask, (void *)priv, 1);
 
+		/* Enable interrupts for all CPUs */
+		mvPp2GbeCpuInterruptsEnable(priv->port, priv->cpuMask);
+
 		/* Unmask Port link interrupt */
 		mvEthPortIsrUnmask(priv->port);
 
@@ -133,7 +136,11 @@ int mv_eth_stop(struct net_device *dev)
 
 	mdelay(10);
 
+	/* Disable interrupts for all CPUs */
+	mvPp2GbeCpuInterruptsDisable(priv->port, priv->cpuMask);
+
 	mv_eth_interrupts_mask(priv);
+	smp_call_function_many(cpu_online_mask, (smp_call_func_t)mv_eth_interrupts_mask, (void *)priv, 1);
 
 	/* make sure that the port finished its Rx polling */
 	for (group = 0; group < MV_ETH_MAX_RXQ; group++)
diff --git a/arch/arm/plat-armada/mv_drivers_lsp/mv_pp2/net_dev/mv_netdev.c b/arch/arm/plat-armada/mv_drivers_lsp/mv_pp2/net_dev/mv_netdev.c
index 037efd6..cdde686 100644
--- a/arch/arm/plat-armada/mv_drivers_lsp/mv_pp2/net_dev/mv_netdev.c
+++ b/arch/arm/plat-armada/mv_drivers_lsp/mv_pp2/net_dev/mv_netdev.c
@@ -314,15 +314,11 @@ int mv_eth_napi_set_cpu_affinity(int port, int group, int cpu_mask)
 		return MV_FAIL;
 	}
 
-	mv_eth_interrupts_mask(pp);
-	smp_call_function_many(cpu_online_mask, (smp_call_func_t)mv_eth_interrupts_mask, (void *)pp, 1);
 	/* update group's CPU mask - remove old CPUs using this group */
 	for_each_possible_cpu(cpu)
 		if ((1 << cpu) & napi_group->cpu_mask)
 			pp->cpu_config[cpu]->napi_group = NULL;
 
-	mvPp2GbeCpuInterruptsDisable(port, napi_group->cpu_mask);
-
 	napi_group->cpu_mask = cpu_mask;
 	napi_group->cause_rx_tx = 0;
 
@@ -330,12 +326,6 @@ int mv_eth_napi_set_cpu_affinity(int port, int group, int cpu_mask)
 		if ((1 << cpu) & cpu_mask)
 			pp->cpu_config[cpu]->napi_group = napi_group;
 
-	mvPp2GbeCpuInterruptsEnable(port, cpu_mask);
-
-	/* mask rx interrupts for every cpu according to its napi group */
-	mv_eth_interrupts_unmask(pp);
-	smp_call_function_many(cpu_online_mask, (smp_call_func_t)mv_eth_interrupts_unmask, (void *)pp, 1);
-
 	return 0;
 }
 
@@ -377,16 +367,9 @@ int mv_eth_napi_set_rxq_affinity(int port, int group, int rxq_mask)
 		return MV_FAIL;
 	}
 
-	mv_eth_interrupts_mask(pp);
-	smp_call_function_many(cpu_online_mask, (smp_call_func_t)mv_eth_interrupts_mask, (void *)pp, 1);
-
 	napi_group->rxq_mask = rxq_mask;
 	napi_group->cause_rx_tx = 0;
 
-	/* mask rx interrupts for every cpu according to its napi group */
-	mv_eth_interrupts_unmask(pp);
-	smp_call_function_many(cpu_online_mask, (smp_call_func_t)mv_eth_interrupts_unmask, (void *)pp, 1);
-
 	return 0;
 }
 
-- 
1.7.5.4

